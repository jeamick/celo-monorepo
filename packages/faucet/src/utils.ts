import Web3 from 'web3'

export function wait(ms: number) {
  return new Promise((resolve) => setTimeout(resolve, ms))
}

export function withTimeLog<A>(name: string, f: (...args: any[]) => Promise<A>) {
  return async (...args: Parameters<typeof f>): Promise<A> => {
    try {
      console.time(name)
      return await f(...args)
    } finally {
      console.timeEnd(name)
    }
  }
}

export async function runWithTimeLog<A>(name: string, f: () => Promise<A>): Promise<A> {
  try {
    console.time(name)
    return await f()
  } finally {
    console.timeEnd(name)
  }
}

/**
 * Generates a temporary account and invite code.
 * TODO(joshua), figure out how to import this from `@celo/utils`
 */
export function generateInviteCode(): {
  address: string
  inviteCode: string
} {
  const web3 = new Web3()
  const tempAccount = web3.eth.accounts.create()
  const address = tempAccount.address
  const temporaryPrivateKey = tempAccount.privateKey
  // Buffer.from doesn't expect a 0x for hex input
  const inviteCode = Buffer.from(temporaryPrivateKey.substring(2), 'hex').toString('base64')
  return { address, inviteCode }
}

// TODO: Copied from @celo/utils, should be removed once usable as a dependency
export function isE164Number(phoneNumber: string) {
  const E164RegEx = /^\+[1-9][0-9]{1,14}$/
  return E164RegEx.test(phoneNumber)
}

// TODO: Copied from @celo/utils, should be removed once usable as a dependency
export const getPhoneHash = (phoneNumber: string): string => {
  if (!phoneNumber || !isE164Number(phoneNumber)) {
    throw Error('Attempting to hash a non-e164 number: ' + phoneNumber)
  }
  const web3 = new Web3()
  return web3.utils.soliditySha3({ type: 'string', value: phoneNumber })
}
