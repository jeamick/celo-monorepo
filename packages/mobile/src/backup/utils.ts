import { generateMnemonic, wordlists } from 'react-native-bip39'

export async function createQuizWordList(mnemonic: string, language: string | null) {
  const disallowedWordSet = new Set(mnemonic.split(' '))
  const languageWordList = getWordlist(language)
  const wordOptions: string = await generateMnemonic(1000, null, languageWordList)
  const quizWordList = new Set(
    [...wordOptions.split(' ')].filter((word: string) => !disallowedWordSet.has(word))
  )
  return [...quizWordList]
}

export function selectQuizWordOptions(correctWord: string, allWords: string[], numOptions: number) {
  const wordOptions = []
  const correctWordPosition = Math.floor(Math.random() * numOptions)
  for (let i = 0; i < numOptions; i++) {
    wordOptions.push(
      i === correctWordPosition
        ? correctWord
        : allWords[Math.floor(Math.random() * allWords.length)]
    )
  }
  return wordOptions
}

export function getWordlist(language: string | null) {
  let wordlist
  switch (language) {
    case 'es': {
      wordlist = wordlists.ES
      break
    }
    default: {
      wordlist = wordlists.EN
    }
  }
  return wordlist
}
