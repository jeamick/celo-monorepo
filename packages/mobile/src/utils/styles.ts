export const colors = {
  GOLD: '#e5b94c',
  GREEN: '#45CD86',
  TEXT_GREY: '#3D3D3D',
  BACKUP_PHRASE_BOX_GREY: '#F0F0F0',
  SELECTED_ANSWER_GREY: '#F6F6F6',
  WORD_OPTION_GREY: '#2E3338',
  OUTLINE_GREY: '#979797',
  LOADING_BAR_GREY: '#DBDBDB',
  BEIGE: '#FFF5E7',
}

export const fontSizes = {
  bodyTextSize: 16,
  titleSize: 36,
}
