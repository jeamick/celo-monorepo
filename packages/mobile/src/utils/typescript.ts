/**
 * Used to assure switch case exhaustiveness in TypeScript
 */
export const assertUnreachable = (x: never): never => {
  throw new Error("Didn't expect to get here")
}

/**
 * Utility type to infer the Props of a ComponentType
 * Inspired by https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react-navigation/index.d.ts
 */
export type InferProps<T extends React.ComponentType<any>> = T extends React.ComponentType<infer P>
  ? P
  : never
