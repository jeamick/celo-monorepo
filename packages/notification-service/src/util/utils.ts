interface ObjectWithStringsAndUndefined {
  [key: string]: string | undefined
}

interface ObjectWithStrings {
  [key: string]: string
}

export function removeEmptyValuesFromObject(obj: ObjectWithStringsAndUndefined) {
  const newObj: ObjectWithStrings = {}
  Object.keys(obj)
    // @ts-ignore
    .filter((k) => obj[k] !== null && obj[k] !== undefined)
    // @ts-ignore
    .forEach((k) => (newObj[k] = obj[k]))
  return newObj
}
