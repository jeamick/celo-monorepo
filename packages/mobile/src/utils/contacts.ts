import { getMinimal, MinimalContact } from 'react-native-contacts'
import { checkContactsPermission } from 'src/utils/androidPermissions'
import Logger from 'src/utils/Logger'

const TAG = 'utils/contacts'

export async function getAllContacts(): Promise<MinimalContact[] | null> {
  return checkContactsPermission().then((contactPermissionsGiven) => {
    if (!contactPermissionsGiven) {
      Logger.warn(TAG, 'Permissions not give for retrieving contacts')
      return null
    }

    return new Promise((resolve, reject) => {
      getMinimal((error, contacts) => {
        if (error) {
          Logger.error(TAG, 'Error getting all contacts', error)
          reject(error)
        }

        if (!contacts) {
          Logger.error(TAG, 'Contacts is null')
          reject('Contacts is null')
        }

        resolve(contacts)
      })
    })
  })
}
