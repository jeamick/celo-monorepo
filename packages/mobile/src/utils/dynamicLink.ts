import firebase from 'react-native-firebase'
import { FIREBASE_WEB_KEY } from 'src/config'
import Logger from 'src/utils/Logger'

const TAG = 'utils/dynamicLink'

export async function dynamicLink(url: string): Promise<string> {
  try {
    // TODO: Move 'celo.page.link' to celo-org-mobile firebase project and then flip this bool once
    // that's done.
    const useRNFirebase = false

    if (useRNFirebase) {
      const link = new firebase.links.DynamicLink(url, 'celo.page.link')

      // TODO: Remove this ts ignore when we upgrade react-native-firebase
      // @ts-ignore: Seems like RN Firebase has the incorrect type for this function call.
      const shortUrl = await firebase.links().createShortDynamicLink(link, 'UNGUESSABLE')
      return shortUrl
    } else {
      const longDynamicLink = `https://celo.page.link/?link=${encodeURIComponent(url)}`
      const apiUrl = `https://firebasedynamiclinks.googleapis.com/v1/shortLinks?key=${FIREBASE_WEB_KEY}`
      const response = await fetch(apiUrl, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({ longDynamicLink }),
      })
      if (response.status !== 200) {
        return url
      }
      const json = await response.json()
      return json.shortLink
    }
  } catch (error) {
    Logger.error(TAG, 'Failed to shorten invite URL: ' + error.toString())
    return url
  }
}
