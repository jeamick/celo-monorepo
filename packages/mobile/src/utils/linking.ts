import { Linking } from 'react-native'
import DeviceInfo from 'react-native-device-info'
import { CELO_VERIFIER_DOWNLOAD_LINK, CELO_VERIFIER_START_MINING_LINK } from 'src/config'
import Logger from 'src/utils/Logger'

export const navigateToVerifierApp = () =>
  navigateToURI({
    uri: CELO_VERIFIER_START_MINING_LINK,
    backupUri: CELO_VERIFIER_DOWNLOAD_LINK,
  })

export function navigateToWalletPlayStorePage() {
  Linking.openURL(`market://details?id=${DeviceInfo.getBundleId()}`)
}

interface NavigateToURIInterface {
  uri: string
  backupUri?: string
}

export const navigateToURI = ({ uri, backupUri }: NavigateToURIInterface) => {
  Linking.canOpenURL(uri).then((supported) => {
    if (supported) {
      Linking.openURL(uri)
    } else {
      Logger.debug('linking/navigateToURI', `Failed to open ${uri}`)
      if (backupUri) {
        Linking.openURL(backupUri)
      }
    }
  })
}
