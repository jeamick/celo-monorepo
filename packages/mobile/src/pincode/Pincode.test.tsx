import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import * as React from 'react'
import { Provider } from 'react-redux'
import * as renderer from 'react-test-renderer'
import Pincode, { Pincode as PincodeUnwrapped } from 'src/pincode/Pincode'
import { createMockStore, getMockI18nProps } from 'test/utils'

describe('Pincode', () => {
  it('renders correctly for education', () => {
    const store = createMockStore()
    const tree = renderer.create(
      <Provider store={store}>
        <Pincode />
      </Provider>
    )
    expect(tree).toMatchSnapshot()
  })

  it('renders correctly for pin enter', () => {
    const wrapper = shallow(
      <PincodeUnwrapped
        showError={jest.fn()}
        hideAlert={jest.fn()}
        pincodeSet={jest.fn()}
        setPin={jest.fn()}
        {...getMockI18nProps()}
      />
    )
    wrapper.find('Button').simulate('press')
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('renders correctly for pin re-enter', () => {
    const wrapper = shallow(
      <PincodeUnwrapped
        showError={jest.fn()}
        hideAlert={jest.fn()}
        pincodeSet={jest.fn()}
        setPin={jest.fn()}
        {...getMockI18nProps()}
      />
    )
    wrapper.find('Button').simulate('press')
    wrapper.find('Button').simulate('press')
    expect(toJson(wrapper)).toMatchSnapshot()
  })
})
