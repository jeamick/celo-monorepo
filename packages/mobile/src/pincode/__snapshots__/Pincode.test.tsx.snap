// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pincode renders correctly for education 1`] = `
<View
  style={
    Object {
      "backgroundColor": "white",
      "flex": 1,
      "justifyContent": "space-between",
    }
  }
>
  <View
    style={
      Object {
        "height": 35,
        "paddingRight": 10,
        "paddingTop": 10,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "width": 70,
        "zIndex": 100,
      }
    }
  >
    <Text
      accessible={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
    >
         
    </Text>
  </View>
  <RCTScrollView>
    <View>
      <View
        style={
          Object {
            "flexDirection": "row",
            "margin": 0,
            "padding": 20,
          }
        }
      >
        <View
          style={
            Object {
              "flex": 1,
            }
          }
        >
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
            testID="CancelPincodeButton"
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#81868B",
                    "fontFamily": "Hind-Regular",
                    "fontSize": 13,
                    "lineHeight": 18,
                  },
                  Object {
                    "color": "#42D689",
                    "fontFamily": "Hind-Medium",
                    "fontSize": 14,
                    "lineHeight": 18,
                  },
                  undefined,
                ]
              }
            >
              cancel
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "paddingLeft": 20,
            "paddingTop": 30,
          }
        }
      >
        <Image
          resizeMode="contain"
          source={
            Object {
              "testUri": "../../../src/images/coins-logo.png",
            }
          }
          style={
            Object {
              "height": 50,
              "width": 50,
            }
          }
        />
      </View>
      <View
        style={
          Object {
            "flex": 1,
            "paddingHorizontal": 25,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontFamily": "Hind-Light",
                "fontSize": 22,
                "paddingBottom": 20,
                "textAlign": "center",
              },
              Object {
                "color": "#2E3338",
                "padding": 25,
                "textAlign": "center",
              },
            ]
          }
          testID="PincodeTitle"
        >
          createPin.title
        </Text>
        <View
          style={
            Object {
              "marginVertical": 10,
              "paddingHorizontal": 20,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#2E3338",
                "fontFamily": "Hind-Regular",
                "fontSize": 16,
                "lineHeight": 24,
              }
            }
          >
            createPin.intro createPin.why
          </Text>
        </View>
        <View
          style={
            Array [
              Object {
                "marginVertical": 10,
                "paddingHorizontal": 20,
              },
            ]
          }
        >
          <Text
            style={
              Object {
                "color": "#2E3338",
                "fontFamily": "Hind-Regular",
                "fontSize": 16,
                "lineHeight": 24,
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#2E3338",
                    "fontFamily": "Hind-Bold",
                    "fontSize": 16,
                    "lineHeight": 24,
                  },
                ]
              }
            >
              important
               
            </Text>
            createPin.warn
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
  <View
    style={
      Object {
        "alignItems": "center",
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderRadius": 3,
            "flexDirection": "row",
          },
          Object {
            "marginVertical": 10,
          },
          Object {
            "marginBottom": 10,
          },
          Object {
            "paddingHorizontal": 20,
            "paddingVertical": 5,
          },
          Object {
            "backgroundColor": "transparent",
          },
        ]
      }
    >
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "transparent",
            "borderColor": "#42D689",
            "borderRadius": 3,
            "borderWidth": 2,
            "flex": 1,
            "height": 50,
            "justifyContent": "center",
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "space-between",
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontFamily": "Hind-SemiBold",
                  "fontSize": 16,
                },
                Object {
                  "color": "#42D689",
                },
                Object {
                  "paddingLeft": 5,
                  "paddingRight": 5,
                },
              ]
            }
          >
            continue
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`Pincode renders correctly for pin enter 1`] = `
<View
  style={
    Object {
      "backgroundColor": "white",
      "flex": 1,
      "justifyContent": "space-between",
    }
  }
>
  <Connect(DevSkipButton)
    nextScreen="RedeemInvite"
  />
  <ScrollViewMock>
    <View
      style={
        Object {
          "flexDirection": "row",
          "margin": 0,
          "padding": 20,
        }
      }
    >
      <View
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <Link
          onPress={[Function]}
          testID="CancelPincodeButton"
        >
          cancel
        </Link>
      </View>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "paddingLeft": 20,
          "paddingTop": 30,
        }
      }
    >
      <Logo
        height={50}
        type="COLOR"
      />
    </View>
    <View
      style={
        Object {
          "flex": 1,
          "paddingHorizontal": 25,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "fontFamily": "Hind-Light",
              "fontSize": 22,
              "paddingBottom": 20,
              "textAlign": "center",
            },
            Object {
              "color": "#2E3338",
              "padding": 25,
              "textAlign": "center",
            },
          ]
        }
      >
        createPin.title
      </Text>
      <NumberInput
        autoFocus={true}
        isSensitiveInput={true}
        keyboardType="numeric"
        onChange={[Function]}
        onSubmit={[Function]}
        placeholder="createPin.yourPin"
        textContentType="password"
        value=""
      />
    </View>
  </ScrollViewMock>
  <View
    style={
      Object {
        "alignItems": "center",
      }
    }
  >
    <Button
      disabled={true}
      onPress={[Function]}
      standard={true}
      style={
        Object {
          "paddingHorizontal": 20,
          "paddingVertical": 5,
        }
      }
      text="continue"
      type="Primary"
    />
  </View>
</View>
`;

exports[`Pincode renders correctly for pin re-enter 1`] = `
<View
  style={
    Object {
      "backgroundColor": "white",
      "flex": 1,
      "justifyContent": "space-between",
    }
  }
>
  <Connect(DevSkipButton)
    nextScreen="RedeemInvite"
  />
  <ScrollViewMock>
    <View
      style={
        Object {
          "flexDirection": "row",
          "margin": 0,
          "padding": 20,
        }
      }
    >
      <View
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <Link
          onPress={[Function]}
          testID="CancelPincodeButton"
        >
          cancel
        </Link>
      </View>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "paddingLeft": 20,
          "paddingTop": 30,
        }
      }
    >
      <Logo
        height={50}
        type="COLOR"
      />
    </View>
    <View
      style={
        Object {
          "flex": 1,
          "paddingHorizontal": 25,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "fontFamily": "Hind-Light",
              "fontSize": 22,
              "paddingBottom": 20,
              "textAlign": "center",
            },
            Object {
              "color": "#2E3338",
              "padding": 25,
              "textAlign": "center",
            },
          ]
        }
      >
        verifyPin.title
      </Text>
      <NumberInput
        autoFocus={true}
        isSensitiveInput={true}
        keyboardType="numeric"
        onChange={[Function]}
        onSubmit={[Function]}
        placeholder="createPin.yourPin"
        textContentType="password"
        value=""
      />
    </View>
  </ScrollViewMock>
  <View
    style={
      Object {
        "alignItems": "center",
      }
    }
  >
    <Button
      disabled={false}
      onPress={[Function]}
      standard={true}
      style={
        Object {
          "paddingHorizontal": 20,
          "paddingVertical": 5,
        }
      }
      text="verifyPin.finalPin"
      type="Primary"
    />
  </View>
</View>
`;
